### ISSUE TITLE
Use hamilton instead of autorest for Azure Container Registry authentication

### ISSUE THREAD

**Issue by simongottschlag** [google/go-containerregistry#1263]
Hi!

Thank you for all the hard work you are doing. :1st_place_medal: 

Right now, `autorest` from Microsoft is used to provide authentication to Azure Container Registry. There's a much smaller project used by the folks over at Hashicorp (used by terraform, azurerm provider, azuread provider) that's much more ... `Go`. And smaller!

I've been testing it out and it seems to be working. Will need a bit of polish but I thought I'd bring up the question here and see what you think about it.

The project is called hamilton and can be found here: https://github.com/manicminer/hamilton

My idea is that we would provide the Azure authentication and ACR token exchange in hamilton and add the helper to `go-containerregistry` (or a separate repository if that's preferred). I think it would be counter productive to remove autorest from the project to add a new one (`github.com/docker/docker-credential-helpers/credentials`) directly after.

We would have to do the following to complete this task:

- [ ] Move usage of autorest from hamilton to separate module: https://github.com/manicminer/hamilton/issues/138
- [ ] Add Azure Container Registry token exchange support: https://github.com/manicminer/hamilton/issues/140
- [ ] Decide where the helper should live  (go-containerregistry or separate repository)
- [x] Create the helper
- [ ] Switch `go-containerregistry` from using `https://github.com/chrismellard/docker-credential-acr-env` to the new helper
- [ ] Profit???

What are your thoughts on this?

**Comment by imjasonh** [comment]
I'm absolutely 💯 in favor of slimming down the dependency graph -- the size of autorest and aws-sdk-go and all the others is a big part of why this repo is split into so many modules, so that the large deps only end up in `k8schain` and `krane`. So any effort to improve that even more would be very welcome!

If you're interested in building another `authn.Keychain` implementation that does ACR, that isn't based on the `acr-env` cred helper, I'd be very interested, and I'd be happy to help however I can.

I don't think it _necessarily_ needs to live inside this repo -- I'm fine having `k8schain` and `krane` depend on that code in whatever repo you put it. Tools like `kaniko` and `cosign` could also switch over.

tl;dr: Sounds great! Where do I sign up? 😄

**Comment by simongottschlag** [comment]
Hi again @imjasonh!

That sounds really good. 😁 I would actually prefer the helper / keychain to be a part of this project instead of somewhere else. If that's OK with you then that's what I'll aim for!

If we add it here, where and how would you want it? A keychain directly in k8schain?

**Comment by imjasonh** [comment]
If possible, I'd love to start with a separate repo and move it in here once we like how it looks, and can get a good feel for how big it'll be. Both in terms of dependency size, and also estimated maintenance burden.

None of the maintainers of this repo have any experience whatsoever with ACR -- as you may be able to tell from all the ACR bugs! 🙃 -- and I'd love to have better e2e tests in place, like we recently got for ECR, to guard against future bugs.

When it's just repackaging other peoples' code it's not so bad (still bugs!), but if we're going to include whole-cloth new code in here I'd love to have better test coverage for it. I don't think any of us even have active Azure accounts. 😆  I don't really have a good answer for that, honestly.

So I think not having to deal with CLAs and code review latency while we iterate on the helper would be nice, and then when it gets into a settled state we can either a.) just reference it from k8schain/krane, or b.) move it in as a separate module, which gets depended on by k8schain/krane.

Does that make sense?

**Comment by simongottschlag** [comment]
Sure thing! I'll make sure to build some unit tests and e2e for it as well.

**Comment by simongottschlag** [comment]
I can't give you any estimates for when it's done, but I'll keep you in the loop here. Will start with finishing up everything in Hamilton first. 😊👍

**Comment by simongottschlag** [comment]
@imjasonh I've been building out the support for ACR in hamilton: https://github.com/manicminer/hamilton/pull/144

We still need to get the autorest dependency moved out of the main module for it to be worth, but I think based on discussions with the maintainer that that it's plausible.

The I've created a separate repo for the helper, still private and will open it up when there's something to show.

**Comment by simongottschlag** [comment]
@imjasonh I've created an initial commit for the new helper: https://github.com/XenitAB/docker-credential-azure

Please take a look at it and all feedback is welcome! The main idea is to use it with `go-containerregistry` and that means I want to change anything according to your thoughts to make it usable.

I've included some E2E tests for it with a container registry on my MSDN for now.

**Comment by imjasonh** [comment]
cc @tcnghia

**Comment by imjasonh** [comment]
This looks really promising! Thanks for working on this ❤️

**Comment by imjasonh** [comment]
So I've tried it out briefly, and mostly it looks good! It's about 1/3rd the size of a dependency as the current acr-env helper. 👍 

A couple API nits:
- `NewACRHelper` takes a `timeout` and a `cache bool`, but it's unclear to me what values I should pass in my case. I don't expect cred generation to take more than a few seconds, and I don't think I want to cache anything, but what do I know. Maybe there should be some default value, and [functional args](https://dave.cheney.net/2014/10/17/functional-options-for-friendly-apis) so I don't have to specify these, with just `acrhelper.New()`. wdyt?
- `Get` should check whether `serverURL` indicates an ACR registry URL, and return an error, like `ecr-login` does: https://github.com/awslabs/amazon-ecr-credential-helper/blob/69c85dc22db6511932bbf119e1a0cc5c90c69a7f/ecr-login/ecr.go#L52 -- if a helper returns any error, `authn.NewKeychainFromHelper` and `authn.NewMultiKeychain` will skip and consult the next keychain in the list.

**Comment by simongottschlag** [comment]
> So I've tried it out briefly, and mostly it looks good! It's about 1/3rd the size of a dependency as the current acr-env helper, so that's great! 🎉 

Awesome!! 🎉



> - `NewACRHelper` takes a `timeout` and a `cache bool`, but it's unclear to me what values I should pass in my case. I don't expect cred generation to take more than a few seconds, and I don't think I want to cache anything, but what do I know. Maybe there should be some default value, and [functional args](https://dave.cheney.net/2014/10/17/functional-options-for-friendly-apis) so I don't have to specify these, with just `acrhelper.New()`. wdyt?

This was actually one part I wasn't sure would be a good addition but thought I start with it to get the conversation started.

The idea is that timeout is used together with the context, making sure that it's possible to configure how long it should try to get a token. This is mainly because the credentials.Helper allow for context propagation.

The cache bool enables us to cache creation of the authorizer (the way we get the Azure token which is used to get the ACR token). This isn't strictly necessary, but makes it possible to get a token at startup and use it going forward and not have to get a new one at first use. Setting it to false disables this behavior.

Functional options is a great addition and I'll get right on it!

I would like to understand:

- Shouldn't we propagate context for everything nowadays?
- Do you think it's a good idea to cache the Azure credentials at startup?
- If we cache them at startup and it fails, should we add a short circuit to `Get()` to fail immediately (since it's most likely not going to be an Azure credential that's being used)?

> - `Get` should check whether `serverURL` indicates an ACR registry URL, and return an error, like `ecr-login` does: https://github.com/awslabs/amazon-ecr-credential-helper/blob/69c85dc22db6511932bbf119e1a0cc5c90c69a7f/ecr-login/ecr.go#L52 -- if a helper returns any error, `authn.NewKeychainFromHelper` and `authn.NewMultiKeychain` will skip and consult the next keychain in the list.

Will add!

Thank you for the feedback! Will try to add it later today.

**Comment by jonjohnsonjr** [comment]
Maybe NewACRHelper should take a context rather than taking a timeout and constructing a timeout?

Is there any downside to the caching? Should it just always happen?

**Comment by simongottschlag** [comment]
@imjasonh functional options is a definite yes. I personally love them, going to add later today if possible. I'm also taking back the idea to short circuit the Get() method, since MSI may take up to 60 seconds to be available in some cases (like using aad-pod-identity) and we don't want to always fail if it failed the first time.

@jonjohnsonjr it would be possible to add context to the constructor, but I expect it could be confusing to use it. I would expect someone to use it with a timeout when constructing it and that would also mean that we shouldn't propagate it to the methods (or at least I've never seen it used that way). Since Get() doesn't support context it is either differently used between construction and execution or it may be interpreted as being propagated (stored in the struct) which seems like a non-Go thing to do.

I'm actually not sure how to handle it in the best way. It would be quite nice if we didn't have to adhere to the interface of credential.Helper..

What are your thoughts on it?

**Comment by simongottschlag** [comment]
@imjasonh I've implemented functional options now. Please take another look and come back with feedback. :+1:

**Comment by imjasonh** [comment]
Looks great to me! I'm trying it out in https://github.com/google/ko/compare/main...imjasonh:include-helpers-acr and the functional args are much nicer. Thanks for making that change!

I don't have an ACR setup handy to test it out, but assuming it works the same as acr-env I'm pretty happy to replace it everywhere in favor of this.

Do you have any timeline in mind for when your hamilton PRs will be upstreamed so we don't need to depend on your fork? Not that it's a blocker, but it'd be nice to have before too long.

**Comment by simongottschlag** [comment]
> 5,945 additions and 18,736 deletions

Nice! 😁👍 What is the current compiled binary size and what's the new one with the change?

Glad you liked the functional args, I do prefer them myself as well!

I'm having a conversation with the hamilton maintainer but it's up him when it comes to the timeline. I'm not in any kind of hurry but I do look forward to them being merged. 😊 I will try to fix any feedback items from him as soon as I can.

I don't mind you using the fork, but if there aren't any issues with autorest I'd wait if possible.

A few things I would like some feedback on:
- the repo name, docker-credential-acr, isn't the greatest. Any better ideas?
- the package name, acrhelper, isn't the best either. Any suggestions?
- is there anything else we can do rather than to use the credentials.Helper interface?

**Comment by imjasonh** [comment]
> Nice! 😁👍 What is the current compiled binary size and what's the new one with the change?

With this change, binary size remains 26 MB, there's no observable increase in binary size 🎉 

> * the repo name, docker-credential-acr, isn't the greatest. Any better ideas?

I don't really have any better alternatives. I'd worry a bit that there might eventually be a "real" ACR cred helper (not creds store) that would be a confusing conflict, but as long as your README points to the other "real" one, that seems fine.

Were you also planning to provide the cred helper as an installable binary, for use with the `docker` CLI? It should be fairly easy to wrap the `Helper` interface to make it runnable, like ecr-login, acr-env, etc. do.

> * the package name, acrhelper, isn't the best either. Any suggestions?

Seems fine to me.

> * is there anything else we can do rather than to use the credentials.Helper interface?

If you want to use `authn.NewKeychainFromHelper` you'd have to use the Helper interface. If you want to build your own `authn.Keychain` implementation just for use in ggcr clients that could be nice, but `Resolve` also doesn't take a context, so it doesn't really help there. 😕

**Comment by imjasonh** [comment]
Hey, gentle ping here. No real rush, just curious what the status is, and if there's anything I can do to help. Not having to depend on all of Azure autorest would be great.

**Comment by simongottschlag** [comment]
Hi @imjasonh! 😊

No worries! There's been progress, but waiting for the following to be merged: https://github.com/manicminer/hamilton/pull/154

When that is done, I just need an review and approval on this PR: https://github.com/manicminer/hamilton/pull/144

As soon as that's done, I'll update the helper and let you know. 😊👍

**Comment by imjasonh** [comment]
Great to hear! I'll watch those PRs then. Excited to see this!

**Comment by simongottschlag** [comment]
Closing this for now since I'm not getting the PR merged in another repo. Will re-open if it comes up again.
